Feat/feature screen one presentation

anarahari opened the pull request.

anarahari committed:
feat : Feature screen one, domain module setup

anarahari committed:
Merge branch 'feat/Workflow_EmptyModules' into feat/Feature_Screen_One_Domain

# Conflicts:
#	common/build.gradle.kts

anarahari committed:
feat : use case, repository, pagination setup

anarahari committed:
Merge branch 'feat/Workflow_EmptyModules' into feat/Feature_Screen_One_Domain

anarahari committed:
feat : setup data layer for screen one

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

# Conflicts:
#	domain/src/main/java/com/compose/domain/paging/CharacterPagingSource.kt
#	domain/src/main/java/com/compose/domain/repository/CharacterRepository.kt
#	domain/src/main/java/com/compose/domain/usecase/GetCharactersUseCase.kt

anarahari committed:
feat : screen one data module setup with graphql, repo impl, dagger modules

anarahari committed:
feat : screen one presentation module setup

anarahari committed:
feat : top app bar for screes

anarahari committed:
feat : top app bar for screens

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
feat : Added datamapper, repositoryimpl changes to integrate data mapper for character model

anarahari committed:
feat : ktlintcheck changes for spaces

anarahari committed:
feat : CommonConstants marked as internal

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Data' into feat/Feature_Screen_One_Presentation

# Conflicts:
#	.idea/deploymentTargetSelector.xml
#	domain/build.gradle.kts

anarahari committed:
feat : FirstScreen topbar, listview using LazyVerticalGrid

anarahari committed:
feat : TopAppBar changes, removed preview function

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -0,0 +1,15 @@
+import com.common.graphql.GetCharactersQuery
+import com.compose.domain.mapper.CharacterModel
+
+fun getCharactersQueryToCharacterModel(results: GetCharactersQuery.Result?): CharacterModel {
+    return CharacterModel(
+        id = results?.character?.id.orEmpty(),
+        name = results?.character?.name.orEmpty(),
+        image = results?.character?.image.orEmpty(),
+        gender = results?.character?.gender.orEmpty(),
+        species = results?.character?.species.orEmpty(),
+        status = results?.character?.status.orEmpty()
+    )
+}
```
Would you want to have a model object if you don't have a results object? What if `results` is null? Would it be better to have a nullable `CharacterModel` in that case?

anarahari replied:
@shubhamsinghmutualmobile In the implementation class we are anyways checking the result object if it is null, it will go to else condition and it will return empty list, so do you think we also need to declare the CharacterModel to be nullable ?

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/paging/CharacterPagingSource.kt:
```diff
@@ -17,19 +16,19 @@ class CharacterPagingSource(
         return try {
-            val response = apolloClient.query(GetCharactersQuery(Optional.present(page))).execute()
-
+            // val response = apolloClient.query(GetCharactersQuery(Optional.present(page))).execute()
```
Let's remove unused code

anarahari replied:
Sure

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/paging/CharacterPagingSource.kt:
```diff
@@ -17,19 +16,19 @@ class CharacterPagingSource(
-                    if (response.data?.characters?.info?.next != null) {
+                val nextKey = page + 1
+                    /*if (response.data?.characters?.info?.next != null) {
                         page + 1
                     } else {
                         null
-                    }
+                    }*/
```
Same

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt:
```diff
@@ -0,0 +1,28 @@
+import kotlinx.coroutines.flow.flow
+import javax.inject.Inject
+
+internal class CharacterRepositoryImpl @Inject constructor(private val apolloClient: ApolloClient) :
+    CharacterRepository {
+    override suspend fun getCharacters(): Flow<List<CharacterModel>> = flow {
+        val response = apolloClient.query(GetCharactersQuery()).execute()
+        if (response.hasErrors()) {
+            emit(emptyList())
+            return@flow
+        }
+        val results = response.data?.characters?.results?.map {
+            getCharactersQueryToCharacterModel(it)
+        } ?: emptyList()
+        emit(results)
+    }.catch {
+        emit(emptyList())
+    }
+}
```
Why do we need a `flow` here when we are not observing/subscribing-to any values?

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt:
```diff
@@ -0,0 +1,104 @@
+import com.compose.presentation.uistate.UiState
+import com.compose.presentation.viewmodel.CharacterViewModel
+
+@Composable
+fun CharacterListScreen(viewModel: CharacterViewModel, modifier: Modifier = Modifier) {
```
Would it be possible to generate a preview of this composable now that you're asking for a viewModel as its param?

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/TopAppBar.kt:
```diff
@@ -0,0 +1,41 @@
+@Composable
+fun CustomAppBar(
+    title: String,
+    backAction: (() -> Unit)? = null
```
General compose param naming conventions recommend that we follow the `on{ACTION}` naming approach for lambdas. In this case, it would be `onBack` or `onBackButtonPressed` depending on how specific you want to be.

anarahari replied:
ok sure

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/TopAppBar.kt:
```diff
@@ -0,0 +1,41 @@
+    backAction: (() -> Unit)? = null
+) {
+    Row(modifier = Modifier.fillMaxWidth()
+        .height(70.dp)
+        .background(MaterialTheme.colorScheme.background),
+        horizontalArrangement = Arrangement.Center,
+        verticalAlignment = Alignment.CenterVertically) {
+        backAction?.let {
+            Box(modifier = Modifier.clickable { backAction }
+                .padding(5.dp)) {
```
Try using the [4-Point Grid System](https://medium.com/@aratidube12lns/4-point-grid-system-for-more-consistent-interface-design-efea81dea3f3) in your designs from now onwards

anarahari replied:
Thanks for sharing
will read and try implementing the same

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/ui/theme/Type.kt:
```diff
@@ -0,0 +1,19 @@
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.sp
+
+val Typography =
```
Better to name these properties based on the project name to easily be able to differentiate between them and the library APIs. For example, this could've been named `RMTypography` (where `RM` means `Rick & Morty`).

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/viewmodel/CharacterViewModel.kt:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    private fun getCharacters() {
+        viewModelScope.launch(Dispatchers.IO) {
```
Instead of using Dispatchers directly, it's better to constructor-inject them to improve the testability of your ViewModel. It would be easier for you to just provide the `StandardTestDispatcher` to it and it will behave appropriately in your tests.

shubhamsinghmutualmobile commented on app/src/main/res/values/themes.xml:
```diff
@@ -1,16 +1,18 @@
-        <item name="android:statusBarColor">?attr/colorPrimaryVariant</item>
-        <!-- Customize your theme here. -->
-    </style>
+<!--    <style name="Theme.RickyMortyPoc" parent="Theme.MaterialComponents.DayNight.DarkActionBar">-->
+<!--        &lt;!&ndash; Primary brand color. &ndash;&gt;-->
+<!--        <item name="colorPrimary">@color/purple_500</item>-->
+<!--        <item name="colorPrimaryVariant">@color/purple_700</item>-->
+<!--        <item name="colorOnPrimary">@color/white</item>-->
+<!--        &lt;!&ndash; Secondary brand color. &ndash;&gt;-->
+<!--        <item name="colorSecondary">@color/teal_200</item>-->
+<!--        <item name="colorSecondaryVariant">@color/teal_700</item>-->
+<!--        <item name="colorOnSecondary">@color/black</item>-->
+<!--        &lt;!&ndash; Status bar color. &ndash;&gt;-->
+<!--        <item name="android:statusBarColor">?attr/colorPrimaryVariant</item>-->
+<!--        &lt;!&ndash; Customize your theme here. &ndash;&gt;-->
+<!--    </style>-->
```
Let's remove all unused code

anarahari replied:
Sure

anarahari committed:
feat : Removed flow from repository impl, added dispatchermodule, removed unwanted codes, rename typography to RMTypography, backactio renamed to onBackButtonPressed

anarahari committed:
feat : dispatcher module injection added in ViewModelModule, ViewModelFactoryProvider, Datamodule

anarahari committed:
feat : ktlint checks

anarahari committed:
feat : data

- CharacterPagingSource removed commented code

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
feat : data

- Changed the implementation type of getCharacters to   List<T> in repository
- Nullable return of charactermodel in datamapper

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Data' into feat/Feature_Screen_One_Presentation

# Conflicts:
#	data/src/main/java/com/compose/data/paging/CharacterPagingSource.kt
#	domain/src/main/java/com/compose/domain/usecase/GetCharactersUseCase.kt

anarahari committed:
feat : presentation

- changes to calling of invoke function in usecase and viewmodel

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/viewmodel/ViewModelModule.kt:
```diff
@@ -0,0 +1,20 @@
+    @Singleton
+    fun provideViewModelFactory(
+        getCharactersUseCase: GetCharactersUseCase,
+        @IoDispatcher ioDispatcher: CoroutineDispatcher
```
Name this (and all other similar params) to `coroutineDispatcher` instead.

anarahari replied:
sure

anarahari committed:
feat : Presentation

Pr reviews changes

- Changed naming convention for ioDispatcher to coroutineDispatcher
